internal/backend/runtime/omni/controllers/omni/redacted_cluster_machine_config.go

```diff
@@ -25,59 +31,269 @@ import (
+	options                 RedactedClusterMachineConfigControllerOptions
+}
 
-				buffer, err := cmc.TypedSpec().Value.GetUncompressedData()
-				if err != nil {
-					return err
-				}
+// RedactedClusterMachineConfigControllerOptions contains options for RedactedClusterMachineConfigController.
+type RedactedClusterMachineConfigControllerOptions struct {
+	Clock               clockwork.Clock
+	DiffCleanupInterval time.Duration
+	DiffMaxAge          time.Duration
+	DiffMaxCount        int
+}
 
-				defer buffer.Free()
+// NewRedactedClusterMachineConfigController creates a new instance of RedactedClusterMachineConfigController with the provided options.
+func NewRedactedClusterMachineConfigController(options RedactedClusterMachineConfigControllerOptions) *RedactedClusterMachineConfigController {
+	if options.DiffCleanupInterval == 0 {
+		options.DiffCleanupInterval = 24 * time.Hour
+	}
 
-				data := buffer.Data()
+	if options.DiffMaxAge == 0 {
+		options.DiffMaxAge = 30 * 24 * time.Hour
+	}
 
-				if data == nil {
-					if err = cmcr.TypedSpec().Value.SetUncompressedData(nil); err != nil {
-						return err
-					}
+	if options.DiffMaxCount == 0 {
+		options.DiffMaxCount = 1000
+	}
 
+	if options.Clock == nil {
+		options.Clock = clockwork.NewRealClock()
+	}
+
+	return &RedactedClusterMachineConfigController{
+		options:                 options,
+		modifiedAtAnnotationKey: omni.SystemLabelPrefix + "modified-at",
+	}
+}
+
+// Name implements controller.QController interface.
+func (ctrl *RedactedClusterMachineConfigController) Name() string {
+	return "RedactedClusterMachineConfigController"
+}
+
+// Settings implements controller.QController interface.
+func (ctrl *RedactedClusterMachineConfigController) Settings() controller.QSettings {
+	return controller.QSettings{
+		RunHook: func(ctx context.Context, logger *zap.Logger, r controller.QRuntime) error {
+			ticker := ctrl.options.Clock.NewTicker(ctrl.options.DiffCleanupInterval)
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
 					return nil
+				case <-ticker.Chan():
+					if err := ctrl.cleanup(ctx, r, logger); err != nil {
+						return err
+					}
 				}
+			}
+		},
+		Inputs: []controller.Input{
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.ClusterMachineConfigType,
+				Kind:      controller.InputQPrimary,
+			},
+		},
+		Outputs: []controller.Output{
+			{
+				Type: omni.RedactedClusterMachineConfigType,
+				Kind: controller.OutputExclusive,
+			},
+			{
+				Type: omni.MachineConfigDiffType,
+				Kind: controller.OutputExclusive,
+			},
+		},
+	}
+}
 
-				config, err := configloader.NewFromBytes(data)
-				if err != nil {
-					return err
-				}
+// Reconcile implements controller.QController interface.
+func (ctrl *RedactedClusterMachineConfigController) Reconcile(ctx context.Context, logger *zap.Logger, r controller.QRuntime, ptr resource.Pointer) error {
+	cmc, err := safe.ReaderGet[*omni.ClusterMachineConfig](ctx, r, ptr)
```

Comment:
I think there should be finalizers on the `ClusterMachineConfig` resource.
It looks like this controller always had this issue.

Reply:
you were right, but the issue was introduced with this PR only, as I converted this controller from qtransform to regular qcontroller. I fixed that now, finalizers are managed correctly.